ui/src/storage/storage.tsx

```diff
@@ -0,0 +1,81 @@
+import Loading from "components/loading";
+import { useAppDispatch } from "hooks";
+import { useCallback } from "react";
+import { useAsync } from "react-async";
+import { AnyAction, Dispatch, Middleware, MiddlewareAPI } from "redux";
+import { loadUserData, RootState } from "rootReducer";
+import * as tanagra from "tanagra-api";
+
+const currentVersion = 1;
+
+export interface StoragePlugin {
+  store(data: tanagra.UserData): Promise<void>;
+  load(): Promise<tanagra.UserData | undefined>;
+}
+
+export const storeUserData: Middleware<unknown, RootState> =
+  (store: MiddlewareAPI<Dispatch<AnyAction>>) =>
+  (next: Dispatch<AnyAction>) =>
+  (action: AnyAction): AnyAction => {
+    const result = next(action);
+    if (storagePlugin) {
+      const state = store.getState();
+
+      const data: tanagra.UserData = {
+        version: currentVersion,
+        cohorts: state.cohorts,
+        conceptSets: state.conceptSets,
+      };
+
+      storagePlugin.store(data);
+    }
+    return result;
+  };
+
+export function LoadingUserData(props: { children?: React.ReactNode }) {
+  const dispatch = useAppDispatch();
+  const status = useAsync<tanagra.UserData>(
+    useCallback(() => {
```

Comment:
Hmm... `useCallback` inlined inside `useAsync`? Is this a typo? Or does this do something that I'm not understanding?